src/control/events/ras.go

```diff
@@ -0,0 +1,167 @@
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package events
+
+/*
+#include "daos_srv/ras.h"
+*/
+import "C"
+
+import (
+	"encoding/json"
+
+	"github.com/pkg/errors"
+
+	mgmtpb "github.com/daos-stack/daos/src/control/common/proto/mgmt"
+)
+
+// Event all custom event types must implement this interface.
+//
+// To/From Proto methods allow conversion regardless of underlying type.
+type Event interface {
+	GetID() RASID
+	GetType() RASTypeID
+	ToProto() (*mgmtpb.RASEvent, error)
+	FromProto(*mgmtpb.RASEvent) error
+}
+
+// NewFromProto creates an Event from given protobuf RASEvent message.
+func NewFromProto(pbEvent *mgmtpb.RASEvent) (Event, error) {
+	switch RASID(pbEvent.Id) {
+	case RASRankExit:
+		rankExit := new(RankExit)
+		return Event(rankExit), rankExit.FromProto(pbEvent)
+	default:
+		return nil, errors.Errorf("unsupported event ID: %d", pbEvent.Id)
+	}
+}
+
+// RASID describes a given RAS event.
+type RASID uint32
+
+// RASID constant definitions.
+const (
+	RASRankExit   RASID = C.RAS_RANK_EXIT
+	RASRankNoResp RASID = C.RAS_RANK_NO_RESP
+)
+
+func (id RASID) String() string {
+	return C.GoString(C.ras_event_id_enum_to_name(uint32(id)))
+}
+
+// Desc returns a description of the event.
+func (id RASID) Desc() string {
+	return C.GoString(C.ras_event_id_enum_to_msg(uint32(id)))
+}
+
+// Uint32 returns uint32 representation of event ID.
+func (id RASID) Uint32() uint32 {
+	return uint32(id)
+}
+
+// RASSeverityID describes the severity of a given RAS event.
+type RASSeverityID uint32
+
+// RASSeverityID constant definitions.
+const (
+	RASSeverityFatal RASSeverityID = C.RAS_SEV_FATAL
+	RASSeverityWarn  RASSeverityID = C.RAS_SEV_WARN
+	RASSeverityError RASSeverityID = C.RAS_SEV_ERROR
+	RASSeverityInfo  RASSeverityID = C.RAS_SEV_INFO
+)
+
+func (sev RASSeverityID) String() string {
+	return C.GoString(C.ras_event_sev_enum_to_name(uint32(sev)))
+}
+
+// Uint32 returns uint32 representation of event severity.
+func (sev RASSeverityID) Uint32() uint32 {
+	return uint32(sev)
+}
+
+// RASTypeID describes the type of a given RAS event.
+type RASTypeID uint32
+
+// RASTypeID constant definitions.
+const (
+	RASTypeAny             RASTypeID = C.RAS_TYPE_ANY
+	RASTypeRankStateChange RASTypeID = C.RAS_TYPE_RANK_STATE_CHANGE
+	RASTypeInfoOnly        RASTypeID = C.RAS_TYPE_INFO_ONLY
+)
+
+func (typ RASTypeID) String() string {
+	return C.GoString(C.ras_event_type_enum_to_name(uint32(typ)))
+}
+
+// Uint32 returns uint32 representation of event type.
+func (typ RASTypeID) Uint32() uint32 {
+	return uint32(typ)
+}
+
+// RASEvent describes details of a specific RAS event.
+type RASEvent struct {
+	Name      string `json:"name"`
+	Timestamp string `json:"timestamp"`
+	Msg       string `json:"msg"`
+	Hostname  string `json:"hostname"`
+	Rank      uint32 `json:"rank"`
+	ID        RASID
+	Severity  RASSeverityID
+	Type      RASTypeID
```

Comment:
In a follow-on PR, I would still like to see these json field names normalized, because these events will be logged and or sent to external consumers. You just need to add the tags here and to the anonymous structs used for the custom marshal/unmarshal. I tested this with some modifications to dump the JSON in `convert.Types()`:

```
data: {
  "id": 1,
  "severity": 3,
  "type": 1,
  "name": "daos_rank_exited",
  "timestamp": "2020-12-17T14:38:47.701802-0500",
  "msg": "DAOS rank exited",
  "hostname": "foo",
  "rank": 1
}
```